[ERM] Unlink linked applications

gherceg opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[JIRA Ticket](https://dimagi-dev.atlassian.net/browse/SS-106)

Keeping this PR low risk by not adding this behavior into the actual delete flow yet.

When a domain link is deleted, any LinkedApplication in the downstream domain should be converted to an Application. This builds on existing copy behavior and deletes the LinkedApplication doc once converted as well.

This could be more sophisticated and only unlink applications with an upstream_id correlating to the upstream domain, but since we do not support a domain having more than 1 upstream domain, it does not seem worth implementing that.

Review by commit.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`LINKED_DOMAINS`
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Wrote tests along the way
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No QA necessary yet.
### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Nothing calls this code besides tests at the moment.
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

gherceg committed:
create method to unlink apps

gherceg committed:
create helper method to fetch all linked apps for a domain

gherceg committed:
create and test method to unlink all linked apps in domain

orangejenny commented on corehq/apps/linked_domain/applications.py:
```diff
@@ -129,6 +139,7 @@ def unlink_app(linked_app):
 
     # remove linked app
-    linked_app.delete()
+    full_linked_app = get_app(linked_app.domain, linked_app._id)
```
We use full apps much more often than briefs. Current naming conventions are that `app` or `linked_app` will refer to an actual app, and then brief apps use `brief` in their identifier.

gherceg replied:
Good to know.

gherceg commented on corehq/apps/linked_domain/applications.py:
```diff
@@ -105,3 +106,19 @@ def link_app(linked_app, master_domain, master_id, remote_details=None):
+    from corehq.apps.app_manager.models import import_app
+    extra_properties = {'name': linked_app.name}
+    app_copy = import_app(linked_app._id, linked_app.domain, extra_properties)
+    app_copy = app_copy.convert_to_application()
```
It was a decision made from ignorance 😄 Thinking back on it, I was under the impression I had to copy the application and then convert the copy, but I see no reason why I can't just convert the existing application like you said. And given that the app id is referenced elsewhere that seems like a much better solution.

gherceg committed:
address Jenny's feedback

- instead of deleting app, convert linked app to app to maintain same app id
- add test case for handling linked and non-linked apps in a domain

gherceg commented on corehq/apps/linked_domain/applications.py:
```diff
@@ -91,6 +91,16 @@ def create_linked_app(master_domain, master_id, target_domain, target_name, remo
     return link_app(linked_app, master_domain, master_id, remote_details)
 
 
+def get_linked_apps_for_domain(domain):
```
Due to a change in `unlink_apps_for_domain` I modified this to fetch full apps. If there are performance concerns, I can change it back to fetch brief apps and convert to full apps when needed.

orangejenny replied:
Sounds good, avoiding optimizing prematurely is good, especially for an infrequent workflow.

gherceg commented on corehq/apps/linked_domain/tests/test_application_util_methods.py:
```diff
@@ -0,0 +1,19 @@
+
+    def test_get_linked_apps_for_domain(self):
+        linked_app1 = LinkedApplication.new_app('domain', 'One')
+        linked_app2 = LinkedApplication.new_app('domain', 'Two')
```
Added an additional test case
